### DISCUSSION TITLE
output.url for RTP streams - jitter

### DISCUSSION THREAD

**Discussion by JonathanThorpe** [savonet/liquidsoap#3684] [User Support]
Hi All,

Running Liquidsoap 2.2.3, I'm attempting to take an SRT stream as an input and send it over RTP to a Barix Extreamer. It appears that the RTP suffers a lot of jitter when using output.url using the following configuration:

```
source_main = input.srt(
  id="srt_main",
  max=10.0,
  clock_safe=false,
  host="192.168.10.25",
  port=8185,
  mode="caller"
)

output.url(url="rtp://192.168.1.97:5000", self_sync=true,
  %ffmpeg(
      format="rtp",
      payload_type=108,
      %audio(codec="pcm_s16le", samplerate=48000, ac=2)
   ), mksafe(source_main))
```

I've tried the following to no avail:
1. `self_sync` as true and false
2. Wrapping the source in a buffer with buffer/max as 1.

If I try the following ffmpeg command (within the LS container), this yields similar results:
```
ffmpeg  -i srt://192.168.10.25:8185 -acodec pcm_s16le -ar 48000 -ac 2 -f rtp -payload_type 108 rtp://192.168.1.97:5000
```

In this instance, I can see that ffmpeg is showing the speed as quite variable - from 1 to 1.2

If I add `-re`, it works fine and shows that the speed is approximately always 1.0.
```
ffmpeg -re -i srt://192.168.10.25:8185 -acodec pcm_s16le -ar 48000 -ac 2 -f rtp -payload_type 108 rtp://192.168.1.97:5000
```

Is there something I can do to yield similar results from Liquidsoap?

**Comment by JonathanThorpe** [comment]
I should note that this works fine but is depending on ffmpeg having the `-re` parameter.

```
output.external(%wav(samplerate=48000, channels=2, samplesize=16, header=false), "ffmpeg -re -f s16le -sample_rate 48000 -ch_layout stereo -i - -acodec pcm_s16le -f rtp -payload_type 108 rtp://192.168.1.97:5000", safe_source,
self_sync=true)
```

There is no explicit mention in the documentation that output.url supports RTP - I assumed it did. If it doesn't, is this the best way to achieve RTP output?

**Comment by toots** [comment]
Thank for you for this detailed report and the comparison with the `ffmpeg` CLI. The issue here is, I believe, a matter of clock. We're trying to maintain a real time ratio of the output stream w.r.t. its content, i.e. that `10s` of content are sent over `10s`. The question being: how to measure that `10s` have passed?

SRT has several modes (see here: https://github.com/Haivision/srt/blob/master/docs/API/API-functions.md#srt_recv). Essentially, in file mode, data is sent as fast as possible but, in live mode, data is sent according to timestamps sent by the sender. Most SRT streams operate in live mode for streaming purposes.

So, first, you should determine if your SRT stream is live or file. If it live, it is usually advisable to keep `clock_safe=true`. This means that we let the `srt` input drive the real time clock, in effect making the sender's clock our real time clock.

If you stream is in file more, however, this is something that we have not tested much. Looking at your description that would be the only assumption that makes sense to explain what you are seeing.

In this mode, the incoming stream comes as fast as possible so the receiver must have its own synchronization mechanism. That's what `-re` does with ffmpeg, using the machine's clock as the real time clock.

I looked at the code and we were missing an option to make this work, which is to set the source as not `self_sync`, which tells the driving clock that the source does not control its latency thus delegating latency control back to the machine's clock. 

I just pushed the changes to add it. Once the build here: https://github.com/savonet/liquidsoap/actions/runs/7728650861 terminates (or a following one if this one fails for some reason), the rolling release for `2.2.x` should be updated with it: https://github.com/savonet/liquidsoap/releases/tag/rolling-release-v2.2.x

I would then recommend to test it with the following options for `input.srt`: `clock_safe=false` and `self_sync=false`. This should put us on-par with what `-re` does in ffmpeg, hopefully making your liquidsoap script work.

**Comment by JonathanThorpe** [comment]
Hi Toots,

Thanks for your fast and detailed reply. I really appreciate the work you do on LiquidSoap.

This should be a live stream coming from another LS - something like this:

```
  enc_tx_1 = %ffmpeg(format = "adts", %audio(codec = "aac", channels = 2, ar = 48000, b="256k"))
  output.srt(max_clients=5, mode="listener", port=config_hls_port_1,
            bind_address=srt_bind,
            enc_tx_1, srt_source_1)

```

When I see a successful build, I'll give it a shot.
